### PR TITLE
feat(api/publish): add new publish management resource for apig api

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:
The APIG services missing resource of the API publishment, user can not pusblish or offline API by terraform.

**Which issue this PR fixes**:
#1594 

**Special notes for your reviewer**:

**Release note**:

```release-note
1. Add new resource of the API publishment and related test and document.
2. Tested some user scenarios:
  a. Publish a new version (without description and the API does not been published).
  b. Publish a new version (without description and the API has been published).
  c. Publish a new version (with description and the API does not been published).
  d. Publish a new version (with description and the API has been published).
  e. Switch a specified version (without description).
  f. Switch a specified version (with description and description is right).
  g. Switch a specified version (with description and description is not right).
  h. Switch a specified version (version does not exist).
  i. Offline current version.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/apig' TESTARGS='-run=TestAccApigApiPublishmentV2_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/apig -v -run=TestAccApigApiPublishmentV2_basic -timeout 360m -parallel 4
=== RUN   TestAccApigApiPublishmentV2_basic
=== PAUSE TestAccApigApiPublishmentV2_basic
=== CONT  TestAccApigApiPublishmentV2_basic
--- PASS: TestAccApigApiPublishmentV2_basic (496.66s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/apig      496.716s
```

### CODE REVIEW

**Comment by Lance52259** [comment]
For this updation,
```
make testacc TEST='./huaweicloud/services/acceptance/apig' TESTARGS='-run=TestAccApigApiPublishmentV2_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/apig -v -run=TestAccApigApiPublishmentV2_basic -timeout 360m -parallel 4
=== RUN   TestAccApigApiPublishmentV2_basic
=== PAUSE TestAccApigApiPublishmentV2_basic
=== CONT  TestAccApigApiPublishmentV2_basic
--- PASS: TestAccApigApiPublishmentV2_basic (501.68s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/apig    501.749s
```

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/apig' TESTARGS='-run=TestAccApigApiPublishmentV2_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/apig -v -run=TestAccApigApiPublishmentV2_basic -timeout 360m -parallel 4
=== RUN   TestAccApigApiPublishmentV2_basic
=== PAUSE TestAccApigApiPublishmentV2_basic
=== CONT  TestAccApigApiPublishmentV2_basic
--- PASS: TestAccApigApiPublishmentV2_basic (499.76s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/apig      499.821s
```

File: docs/resources/apig_api_publishment.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+* `region` - (Optional, String, ForceNew) Specifies the region in which to publish API.
+  If omitted, the provider-level region will be used. Changing this will create a new publishment resource.
+
+* `instance_id` - (Required, String, ForceNew) Specifies an ID of the APIG dedicated instance to which the API belongs
+  to. Changing this will create a new publishment resource.
+
+* `env_id` - (Required, String, ForceNew) Specifies the environment ID to which the current version of the API is published.
+
+* `api_id` - (Required, String, ForceNew) Specifies the API that.
```
Comment (hmbig2) [inline docs/resources/apig_api_publishment.md]: It seems that something is missing behind `that`

File: huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+	return nil
+}
+
+// ResourceApigApiPublishmentCreate is a function that uses terraform configuration to publish new versions or switch
+// historical versions.
+func ResourceApigApiPublishmentCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*config.Config)
+	c, err := config.ApigV2Client(config.GetRegion(d))
+	if err != nil {
+		return fmtp.DiagErrorf("Error creating HuaweiCloud compute client: %s", err)
```
Comment (hmbig2) [inline huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go]: actual it is apigV2Client

File: docs/resources/apig_api_publishment.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+The `histories` block supports:
+
+* `version_id` - Version ID of the API publishment.
+
+* `description` - Version description of the API publishment.
+
+## Import
+
+APIs can be imported using their `instance_id`, `env_id` and `api_id`, separated by colons, e.g.
```
Comment (ShiChangkuo) [inline docs/resources/apig_api_publishment.md]: please seprated them with /

File: huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+func isPublished(versionList []apis.ApiVersionInfo, versionId string) (apis.ApiVersionInfo, bool) {
+	for _, ver := range versionList {
+		if ver.VersionId == versionId {
+			return ver, true
+		}
+	}
+	return apis.ApiVersionInfo{}, false
+}
+
+func isNewestVersion(versionList []apis.ApiVersionInfo, versionId string) (apis.ApiVersionInfo, bool) {
```
Comment (ShiChangkuo) [inline huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go]: suggest to ues isLatestVersion

File: huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+	}
+
+	instanceId := d.Get("instance_id").(string)
+	apiId := d.Get("api_id").(string)
+	envId := d.Get("env_id").(string)
+	// If version_id is configured, try to switch to the corresponding version.
+	if versionId, hasVer := d.GetOk("version_id"); hasVer {
+		histories, err := GetVersionHistories(c, instanceId, envId, apiId)
+		if err != nil {
+			return fmtp.DiagErrorf("Error finding the publish versions of the API (%s) in the environment(%s): %s", err)
```
Comment (ShiChangkuo) [inline huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go]: seems that some parameters are missing

File: huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+}
+
+// GetVersionHistories is a function that obtains version histories by APIG IDs.
+func GetVersionHistories(c *golangsdk.ServiceClient, instanceId, envId, apiId string) ([]apis.ApiVersionInfo, error) {
+	// Get version histories of the API from the specified environment.
+	pages, err := apis.ListPublishHistories(c, instanceId, apiId, apis.ListPublishHistoriesOpts{
+		EnvId: envId,
+	}).AllPages()
+	if err != nil {
+		return []apis.ApiVersionInfo{}, fmtp.Errorf("Error retrieving publish histories: %s", err)
```
Comment (ShiChangkuo) [inline huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go]: suggest to return (nil, error) when we got an error

File: huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+	}
+	histories, err := apis.ExtractHistories(pages)
+	if err != nil {
+		return []apis.ApiVersionInfo{}, fmtp.Errorf("Error extracting publish histories: %s", err)
```
Comment (ShiChangkuo) [inline huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go]: ditto

File: docs/resources/apig_api_publishment.md (hunk 1 above)
Comment (Lance52259) [reply docs/resources/apig_api_publishment.md]: The description of `api_id` has been added.

File: huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go (hunk 2 above)
Comment (Lance52259) [reply huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go]: fixed

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/apig' TESTARGS='-run=TestAccApigApiPublishmentV2_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/apig -v -run=TestAccApigApiPublishmentV2_basic -timeout 360m -parallel 4
=== RUN   TestAccApigApiPublishmentV2_basic
=== PAUSE TestAccApigApiPublishmentV2_basic
=== CONT  TestAccApigApiPublishmentV2_basic
--- PASS: TestAccApigApiPublishmentV2_basic (508.87s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/apig      508.974s
```

File: huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go (hunk 6 above)
Comment (Lance52259) [reply huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go]: sure, for this return and others of the same kind, a nil object or pointer is returned.

File: huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go (hunk 7 above)
Comment (Lance52259) [reply huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go]: done

File: huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go (hunk 5 above)
Comment (Lance52259) [reply huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go]: The api ID and environment ID has been added.

File: docs/resources/apig_api_publishment.md (hunk 3 above)
Comment (Lance52259) [reply docs/resources/apig_api_publishment.md]: The description has been updated.

File: huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go (hunk 4 above)
Comment (Lance52259) [reply huaweicloud/services/apig/resource_huaweicloud_apig_api_publishment.go]: It seems be better.

**Pull request merged**